sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +380,487 @@ def transform(self, X):
+                                        axis=0)
+        X_filled = self.initial_imputer_.fit_transform(X)
+
+        valid_mask = np.logical_not(np.isnan(
+            self.initial_imputer_.statistics_
+        ))
+        self._valid_statistics_indexes = np.flatnonzero(valid_mask)
+        X = X[:, self._valid_statistics_indexes]
+        mask_missing_values = mask_missing_values[
+                              :,
+                              self._valid_statistics_indexes
+        ]
+
+        # perform imputations
+        self._trained_estimator_triplets = []
+        n_samples, n_features = X_filled.shape
+        total_rounds = self.n_burn_in + self.n_imputations
+        results_list = []
+        if self.verbose:
+            print("[MICE] Completing matrix with shape %s" % (X.shape,))
+            start_t = time()
+            mice_msg = '[MICE] Ending imputation round '
+        for iter in range(total_rounds):
+            # order in which to impute
+            ordered_idx = self._get_ordered_idx(mask_missing_values)
+
+            # abs_correlation matrix is used to choose a subset of other
+            # features to impute from
+            abs_corr_mat = self._get_abs_correlation_matrix(X_filled)
+
+            # Fill in each feature in the order of ordered_idx
+            for feat_idx in ordered_idx:
+                neighbor_feat_idx = self._get_neighbor_feat_idx(
+                    n_features,
+                    feat_idx,
+                    abs_corr_mat
+                )
+                X_filled, estimator = self._impute_one_feature(
+                    X_filled,
+                    mask_missing_values,
+                    feat_idx,
+                    neighbor_feat_idx
+                )
+                estimator_triplet = (
+                    feat_idx,
+                    neighbor_feat_idx,
+                    estimator
+                )
+                self._trained_estimator_triplets.append(estimator_triplet)
+
+            if iter >= self.n_burn_in:
+                results_list.append(X_filled[mask_missing_values])
+            if self.verbose:
+                print(mice_msg + 'round %d/%d, elapsed time %0.2f'
+                      % (iter + 1, total_rounds, time() - start_t))
+
+        if len(results_list) > 0:
+            X[mask_missing_values] = np.array(results_list).mean(axis=0)
+        else:
+            X[mask_missing_values] = X_filled[mask_missing_values]
+
+        return X
+
+    def transform(self, X):
+        """Imputes all missing values in X.
+
+        Parameters
+        ----------
+        X : array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        Returns
+        -------
+        X : array-like, shape (n_samples, n_features)
+            The imputed input data.
+        """
+        check_is_fitted(self, 'initial_imputer_')
+        X = check_array(X, dtype=np.float64, force_all_finite=False)
```

Comment:
coming on factorizing fit_transform and transform, you can create a private method `_initial_imputation` containing

```python
        X = check_array(X, dtype=np.float32, order="F", force_all_finite=False)

        # initial imputation
        mask_missing_values = _get_mask(X, self.missing_values)
        self.initial_imputer_ = Imputer(missing_values=self.missing_values,
                                        strategy=self.initial_strategy,
                                        axis=0)
        X_filled = self.initial_imputer_.fit_transform(X)

        valid_mask = np.flatnonzero(np.logical_not(
                                  np.isnan(self.initial_imputer_.statistics_)))
        X = X[:, valid_mask]
        mask_missing_values = mask_missing_values[:, valid_mask]
```

Reply:
Don't we need to store the `Imputer`, to use in `Transform` after `fit`, and not refit every time?

Reply:
I think this would work:

```
        X = check_array(X, dtype=np.float32, order="F", force_all_finite=False)

        # initial imputation
        mask_missing_values = _get_mask(X, self.missing_values)
        if self.initial_imputer_ is None:
            self.initial_imputer_ = Imputer(missing_values=self.missing_values,
                                            strategy=self.initial_strategy,
                                            axis=0)
        X_filled = self.initial_imputer_.fit_transform(X)

        valid_mask = np.flatnonzero(np.logical_not(
            np.isnan(self.initial_imputer_.statistics_))
        )
        X = X[:, valid_mask]
        mask_missing_values = mask_missing_values[:, valid_mask]
```